File: src/musikdroid/app/src/main/java/io/casey/musikcube/remote/playback/PlaybackService.java
package io.casey.musikcube.remote.playback;

import io.casey.musikcube.remote.ui.model.TrackListSlidingWindow;

public interface PlaybackService {
    interface EventListener {
        void onStateUpdated();
    }

    void connect(final EventListener listener);
    void disconnect(final EventListener listener);

    void playAll();

    void playAll(final int index, final String filter);

    void play(
        final String category,
        final long categoryId,
        final int index,
        final String filter);

    void playAt(final int index);

    void pauseOrResume();

    void pause();
    void resume();
    void prev();
    void next();
    void stop();

    void volumeUp();
    void volumeDown();

    void seekForward();
    void seekBackward();
    void seekTo(double seconds);

    int getQueueCount();
    int getQueuePosition();

    double getVolume();
    double getDuration();
    double getCurrentTime();

    PlaybackState getPlaybackState();

    void toggleShuffle();
    boolean isShuffled();

    void toggleMute();
    boolean isMuted();

    void toggleRepeatMode();
    RepeatMode getRepeatMode();

    TrackListSlidingWindow.QueryFactory getPlaylistQueryFactory();

    String getTrackString(final String key, final String defaultValue);
    long getTrackLong(final String key, final long defaultValue);
}

File: src/musikdroid/app/src/main/java/io/casey/musikcube/remote/playback/PlaybackService.kt
package io.casey.musikcube.remote.playback

import io.casey.musikcube.remote.ui.model.TrackListSlidingWindow

interface PlaybackService {
    fun connect(listener: () -> Unit)
    fun disconnect(listener: () -> Unit)

    fun playAll()

    fun playAll(index: Int, filter: String)

    fun play(
        category: String,
        categoryId: Long,
        index: Int,
        filter: String)

    fun playAt(index: Int)

    fun pauseOrResume()

    fun pause()
    fun resume()
    fun prev()
    fun next()
    fun stop()

    fun volumeUp()
    fun volumeDown()

    fun seekForward()
    fun seekBackward()
    fun seekTo(seconds: Double)

    val queueCount: Int
    val queuePosition: Int

    val volume: Double
    val duration: Double
    val currentTime: Double
    val bufferedTime: Double

    val playbackState: PlaybackState

    fun toggleShuffle()
    val isShuffled: Boolean

    fun toggleMute()
    val isMuted: Boolean

    fun toggleRepeatMode()
    val repeatMode: RepeatMode

    val playlistQueryFactory: TrackListSlidingWindow.QueryFactory

    fun getTrackString(key: String, defaultValue: String): String
    fun getTrackLong(key: String, defaultValue: Long): Long
}
